test/circuit.js

```diff
@@ -0,0 +1,290 @@
+/* eslint max-nested-callbacks: ["error", 8] */
+/* eslint-env mocha */
+'use strict'
+
+const chai = require('chai')
+const dirtyChai = require('dirty-chai')
+const expect = chai.expect
+chai.use(dirtyChai)
+
+const parallel = require('async/parallel')
+const series = require('async/series')
+
+const isNode = require('detect-node')
+
+const utils = require('./utils/circuit')
+
+const proc = utils.setUpProcNode
+const js = utils.setUpJsNode
+const go = utils.setUpGoNode
+
+const ws = utils.wsAddr
+const star = utils.wsStarAddr
+const circuit = utils.circuitAddr
+
+const send = utils.send
+
+const base = '/ip4/127.0.0.1/tcp/0'
+
+const connect = (nodeA, nodeB, relay, timeout, callback) => {
+  if (typeof timeout === 'function') {
+    callback = timeout
+    timeout = 1000
+  }
+
+  series([
+    (cb) => nodeA.ipfsd.api.swarm.connect(ws(relay.addrs), cb),
+    (cb) => setTimeout(cb, timeout),
+    (cb) => nodeB.ipfsd.api.swarm.connect(ws(relay.addrs), cb),
+    (cb) => setTimeout(cb, timeout),
+    (cb) => nodeA.ipfsd.api.swarm.connect(circuit(nodeB.addrs), cb)
+  ], callback)
+}
+
+const timeout = 80 * 1000
+const baseTest = {
+  connect,
+  send,
+  timeout
+}
+
+let tests = {
+  'go-go-go': {
+    create: (callback) => series([
+      (cb) => go([`${base}/ws`], cb),
+      (cb) => go([`${base}/ws`], cb),
+      (cb) => go([`${base}/ws`], cb)
+    ], callback)
+  },
+  'js-go-go': {
+    create: (callback) => series([
+      (cb) => js([`${base}/ws`], cb),
+      (cb) => go([`${base}/ws`], cb),
+      (cb) => go([`${base}/ws`], cb)
+    ], callback)
+  },
+  'go-go-js': {
+    create: (callback) => series([
+      (cb) => go([`${base}/ws`], cb),
+      (cb) => go([`${base}/ws`], cb),
+      (cb) => js([`${base}/ws`], cb)
+    ], callback)
+  },
+  'js-go-js': {
+    create: (callback) => series([
+      (cb) => js([`${base}/ws`], cb),
+      (cb) => go([`${base}/ws`], cb),
+      (cb) => js([`${base}/ws`], cb)
+    ], callback)
+  },
+  'go-js-go': {
+    create: (callback) => series([
+      (cb) => go([`${base}/ws`], cb),
+      (cb) => js([`${base}/ws`], cb),
+      (cb) => go([`${base}/ws`], cb)
+    ], callback)
+  },
+  'js-js-go': {
+    create: (callback) => series([
+      (cb) => js([`${base}/ws`], cb),
+      (cb) => js([`${base}/ws`], cb),
+      (cb) => go([`${base}/ws`], cb)
+    ], callback)
+  },
+  'go-js-js': {
+    create: (callback) => series([
+      (cb) => go([`${base}/ws`], cb),
+      (cb) => js([`${base}/ws`], cb),
+      (cb) => js([`${base}/ws`], cb)
+    ], callback)
+  },
+  'js-js-js': {
+    create: (callback) => series([
+      (cb) => js([`${base}/ws`], cb),
+      (cb) => js([`${base}/ws`], cb),
+      (cb) => js([`${base}/ws`], cb)
+    ], callback)
+  }
+}
+
+const connWithTimeout = (timeout) => {
+  return (nodeA, nodeB, relay, callback) => {
+    connect(nodeA, nodeB, relay, timeout, callback)
+  }
+}
+const browser = {
+  'browser-go-js': {
```

Comment:
Keeping track of these mutations for a human is not super easy. Maybe we could generate the combinations given a list of implementations to test?

For example, passing `['js', 'go', 'browser']` to a function that can figure out all permutations of those elements. Would also be really useful to have a parameter that would output combinations of two elements rather than three (circuit needs three elements, middle is relay, but for the rests of interop tests, we only need to elements connecting to each other).

Might be a bit out-of-scope for this PR though.

Reply:
Yeah, I like pre-generating as well. It gets a little tricky because of how nodes connect - `nodes to relay or relay to node`. I took a stab at a more DSLish approach here - https://github.com/ipfs/interop/pull/6/commits/0af7ac56fdd1677d5c0e326b2b9f4bd1cc81ee20#diff-3049b1205b403a9fbacb7734aa167858, which demostrates the amount of data I had to encode in order to perform the spawning and connecting the nodes. I'll play around with it a bit more and I'm sure we can clean this up further, but I think this is "good enough" for now.